Message queue support LIFO

TaiJuWu opened the pull request:
The logic is same with `k_msgq_put_front`.
The main difference is `k_msgq_put_front` is moving `read_ptr` of msgq.
When users use `k_msgq_put_front`, the new data is put at the `read_ptr` so
it will be got if users call `z_impl_k_msgq_get`

resolve:  #46339

cfriedt reviewed (commented):
It looks like put() and put_front() could maybe be merged into the same function.

Also, it appears that the function can return in several locations without calling `k_spin_unlock()`. That should probably be corrected.

`K_SPINLOCK()` is a nice utility that helps enforce correct locking and unlocking.

peter-mitsis commented on kernel/msg_q.c:
```diff
@@ -195,6 +197,83 @@ static inline int z_vrfy_k_msgq_put(struct k_msgq *msgq, const void *data,
+
+	key = k_spin_lock(&msgq->lock);
+
+	SYS_PORT_TRACING_OBJ_FUNC_ENTER(k_msgq, put, msgq, timeout);
```
It might be a good idea to update the parameters to SYS_PORT_TRACING_OBJ_FUNC_xxx() used in this routine.

TaiJuWu replied:
Done.

peter-mitsis commented on kernel/msg_q.c:
```diff
@@ -195,6 +197,83 @@ static inline int z_vrfy_k_msgq_put(struct k_msgq *msgq, const void *data,
+		/* don't wait for message space to become available */
+		result = -ENOMSG;
+	} else {
+		SYS_PORT_TRACING_OBJ_FUNC_BLOCKING(k_msgq, put, msgq, timeout);
+
+		/* wait for put message success, failure, or timeout */
+		_current->base.swap_data = (void *) data;
+
+		result = z_pend_curr(&msgq->lock, key, &msgq->wait_q, timeout);
+		SYS_PORT_TRACING_OBJ_FUNC_EXIT(k_msgq, put, msgq, timeout, result);
+		return result;
```
I think that allowing the caller to pend with a timeout can lead to inconsistent behaviour. That is, if the msgq buffer is not full, then the message will be copied into the front of the buffer. However, if the msgq buffer is full, then the message is being copied into the end of the buffer when we get a message from the msgq (and subsequently remove it from the wait queue).  Because of this, I am wondering if the timeout parameter ought to be removed completely, and k_msgq_put_front() be treated as if it had a timeout of K_NO_WAIT.

TaiJuWu replied:
> I think that allowing the caller to pend with a timeout can lead to inconsistent behaviour. 

Sorry, I don't get the point.

> That is, if the msgq buffer is not full, then the message will be copied into the front of the buffer. 
However, if the msgq buffer is full, then the message is being copied into the end of the buffer when we get a message from the msgq (and subsequently remove it from the wait queue). 

The behavior is correct.

> Because of this, I am wondering if the timeout parameter ought to be removed completely, and k_msgq_put_front() be treated as if it had a timeout of K_NO_WAIT.

I don't get the point, too. Could you explain more detail?

peter-mitsis replied:
There is an implicit property being associated with the message--the message should take priority and be placed at the front of the message queue. As long as the message queue buffer is not full (and thus there is space to put this prioritized message) this works correctly.

My concern is that when the message queue buffer is full and the caller has to pend, then that implicit property is lost and the caller has no indication that this happened or could happen.

Imagine the following scenario ...

The message queue buffer is full. Thread A calls k_msgq_put_front(), but since the buffer is full, Thread A pends. Thread B comes along and calls k_msgq_get(). It does not get message A; it gets message X--the first in the message queue buffer. Internally, k_msgq_get() will then unpend the first waiting thread (which could be Thread A) and copy its message to the end of the message queue buffer. The implicit property of message A is lost, and there is no indication to the caller of k_msgq_put_front() that the message was copied to an unexpected location.

TaiJuWu replied:
> Imagine the following scenario ...
> 
> The message queue buffer is full. Thread A calls k_msgq_put_front(), but since the buffer is full, Thread A pends. Thread B comes along and calls k_msgq_get(). It does not get message A; it gets message X--the first in the message queue buffer. Internally, k_msgq_get() will then unpend the first waiting thread (which could be Thread A) and copy its message to the end of the message queue buffer. The implicit property of message A is lost, and there is no indication to the caller of k_msgq_put_front() that the message was copied to an unexpected location.

You are right!!!!
When `thread A` wake up, it will copy data to `write_ptr` but In my expectation, it should move `read_ptr` and then copy.
The main reason is we don't record the reason make `thread A` sleep.
We should record the pending reason is `k_msgq_put` or `k_msgq_put_front`

But  there is no bit can be used in `thread_state`.
Do you have any suggestion?

TaiJuWu replied:
> There is an implicit property being associated with the message--the message should take priority and be placed at the front of the message queue. As long as the message queue buffer is not full (and thus there is space to put this prioritized message) this works correctly.
> 
> My concern is that when the message queue buffer is full and the caller has to pend, then that implicit property is lost and the caller has no indication that this happened or could happen.
> 
> Imagine the following scenario ...
> 
> The message queue buffer is full. Thread A calls k_msgq_put_front(), but since the buffer is full, Thread A pends. Thread B comes along and calls k_msgq_get(). It does not get message A; it gets message X--the first in the message queue buffer. Internally, k_msgq_get() will then unpend the first waiting thread (which could be Thread A) and copy its message to the end of the message queue buffer. The implicit property of message A is lost, and there is no indication to the caller of k_msgq_put_front() that the message was copied to an unexpected location.

You are right!!!!

When thread A wake up, it will copy data to write_ptr but In my expectation, it should move read_ptr and then copy.

The main reason is we don't record the reason make thread A sleep.
We should record the pending reason is k_msgq_put or k_msgq_put_front and when thread A unpend, it have to execute the corresponding operation.

But there is no bit can be used in thread_state.
Do you have any suggestion?

andyross commented on kernel/msg_q.c:
```diff
@@ -195,6 +197,83 @@ static inline int z_vrfy_k_msgq_put(struct k_msgq *msgq, const void *data,
+
+	SYS_PORT_TRACING_OBJ_FUNC_ENTER(k_msgq, put, msgq, timeout);
+
+	if (msgq->used_msgs < msgq->max_msgs) {
```
Would **strongly** prefer this logic be unified with the rest of the file instead of being a cut/paste "mostly identical"  version, i.e. put the "empty queue == deliver synchronously" in a utility shared by the other code, etc...

TaiJuWu replied:
Done.

andyross commented on include/zephyr/kernel.h:
```diff
@@ -4569,6 +4569,29 @@ int k_msgq_cleanup(struct k_msgq *msgq);
+/**
+ * @brief Send a message to front of a message queue.
+ *
+ * This routine sends a message to message queue @a q.
```
Docs don't actually say what the function does, should be something like: "This routine sends a message to message queue q, inserting it before any other messages already in the queue."

TaiJuWu replied:
Done.

TaiJuWu commented:
> It looks like put() and put_front() could maybe be merged into the same function.

Done.

TaiJuWu commented:
> It looks like put() and put_front() could maybe be merged into the same function.

Done.

cfriedt commented on kernel/msg_q.c:
```diff
@@ -166,23 +178,29 @@ int z_impl_k_msgq_put(struct k_msgq *msgq, const void *data, k_timeout_t timeout
 
+int z_impl_k_msgq_put(struct k_msgq *msgq, const void *data, k_timeout_t timeout)
+{
+	return z_impl_k_msgq_put_internal(msgq, data, timeout, false);
```
I would guess that the sys_trace macros should probably go in the wrapper functions, because those are the externally visible API calls that people will expect to trace.

TaiJuWu replied:
Do you mean add a api called `sys_port_trace_k_msgq_put_front_blocking` to `z_impl_k_msgq_put_front`?
If so, there is five lines need to be modified for `SYS_PORT_TRACING_OBJ_FUNC_XXX` and these lines is distributed in `z_impl_k_msgq_put`.

There is a question which is I don't have good ideal to deal the same logic in `z_impl_k_msgq_put` and `z_impl_k_msgq_put_front`.

andyross replied:
(I take no position on tracing.  Honestly I think our current tracing layer is sort of a kludge and constantly breaking as code gets moved around.  Tracing done well needs to happen at a lower level, like automatic toolchain annotation or dtrace-style dynamic hooking.)

nashif replied:
> Do you mean add a api called `sys_port_trace_k_msgq_put_front_blocking` to `z_impl_k_msgq_put_front`?
> If so, there are five lines need to be modified for `SYS_PORT_TRACING_OBJ_FUNC_XXX` and these lines is distributed in `z_impl_k_msgq_put`.

The call to the tracing macros need to reference the the action being called (put or prepend), so in case of put you will be calling `SYS_PORT_TRACING_OBJ_FUNC_ENTER(k_msgq, put, msgq, timeout);` for example, in the case of prepending, you will call `SYS_PORT_TRACING_OBJ_FUNC_ENTER(k_msgq, prepend, msgq, timeout);`

teburd replied:
> tracing

I briefly looked at how another rtos does dynamic probes but it it looks like at least in that instance it'd require setting hardware break points and triggering NMIs, and deciphering some state of where the program was before the interrupt, recording it, and returning. Not saying its not clever or useful but compared to the macros we have today... pretty heavy handed in terms of over head.

I *do* love the idea that we could arbitrarily add probes like that though, its clever and cool for sure.

TaiJuWu committed:
kernel: introduce k_msgq_prepend api

The logic is same with `k_msgq_prepend`.
The main difference is `k_msgq_prepend` is moving `read_ptr` of msgq.
When users use `k_msgq_prepend`, the new data is put at the `read_ptr` so
it will be got if users call `z_impl_k_msgq_get`

Signed-off-by: TaiJu Wu <tjwu1217@gmail.com>

andyross reviewed (dismissed):
This looks correct to me.

nashif commented on include/zephyr/kernel.h:
```diff
@@ -4569,6 +4570,30 @@ int k_msgq_cleanup(struct k_msgq *msgq);
+ * @retval -ENOMSG Returned without waiting or queue purged.
+ * @retval -EAGAIN Waiting period timed out.
+ */
+__syscall int k_msgq_put_front(struct k_msgq *msgq, const void *data, k_timeout_t timeout);
```
maybe k_msgq_prepend

nashif commented on kernel/msg_q.c:
```diff
@@ -121,7 +121,8 @@ int k_msgq_cleanup(struct k_msgq *msgq)
 
 
-int z_impl_k_msgq_put(struct k_msgq *msgq, const void *data, k_timeout_t timeout)
+static int z_impl_k_msgq_put_internal(struct k_msgq *msgq, const void *data,
```
you should remove the  `z_impl_` prefix here, this should only be used for syscalls.

nashif commented:
additionally, a test is needed.

TaiJuWu commented:
Hi @nashif @peter-mitsis @andyross @cfriedt ,

There are two problems about this PR, hope you can give me suggestion.

1. Like @peter-mitsis said, [here](https://github.com/zephyrproject-rtos/zephyr/pull/71195#discussion_r1562833874), there is a problem about `msgq_put_front`. There are two proposal.
  
    a. Like @peter-mitsis's suggestion, `msgq_put_front ` does not be allowed to use `timeout`, it always use `K_NO_WAIT` and documented.
  
      b. Add a flag for `k_thread_t`, this flag can be used to point the caller call `msgq_put_front` or `msgq_put` and when the thread is unpended, we know reason of thread pending.
  The other bits of this flag can be reserved. For example, we can point this thread is deadline task or normal task.
  
 2. Tracing, You can reference [here](https://github.com/zephyrproject-rtos/zephyr/pull/71195#discussion_r1561871283).:
    a. Split tracing for `msgq_put_front ` and `msgq_put` but there are many similar code.

    b. Use common logic but only one tracing for `msgq_put_front ` and `msgq_put`.

   
   
For question 1, I prefer to add an additional flag to deal this problem.
For question 2, I take no position like @andyross .

If you or anyone have any additional suggestion, please tell me, thanks a lot.

cfriedt commented:
@TaiJuWu - I think the tracing calls can go into one function, but they likely do need names conditional on the entry points, as @nashif also said (regardless of how well the tracing API works).

Originally, I was thinking "prepend" instead of "put_front", and "append" instead of "put" but then that would be a breaking change. To me the more minimal and intuitive change is better. Also, naming things is hard 😅. My input is probably useless on that.

Otherwise, I think that would do it for me.

TaiJuWu commented:
> I think that would do it for me.

Thank for your reply. Your suggestion is also import for me.

TaiJuWu commented:
> @TaiJuWu - I think the tracing calls can go into one function, but they likely do need names conditional on the entry points, as @nashif also said (regardless of how well the tracing API works).
> 
> Originally, I was thinking "prepend" instead of "put_front", and "append" instead of "put" but then that would be a breaking change. To me the more minimal and intuitive change is better. Also, naming things is hard 😅. My input is probably useless on that.
> 
> Otherwise, I think that would do it for me.

Thank for your reply. Your suggestion is also import for me.

cfriedt commented:
@TaiJuWu - I think something like this should work

```cpp
if (port_put_front) {
  SYS_PORT_TRACING_OBJ_FUNC_EXIT(k_msgq, put, msgq, timeout, 0);
} else {
SYS_PORT_TRACING_OBJ_FUNC_EXIT(k_msgq, put_back, msgq, timeout, 0);
}
```

TaiJuWu committed:
test: add config to test msgq  prepend featrue

Add a config `CONFIG_TEST_MSGQ_PREPEND` in mesgq_test to
enable message queue prepend feature.

Signed-off-by: TaiJu Wu <tjwu1217@gmail.com>

peter-mitsis commented on kernel/msg_q.c:
```diff
@@ -239,14 +280,25 @@ int z_impl_k_msgq_get(struct k_msgq *msgq, void *data, k_timeout_t timeout)
+				if (msgq->read_ptr == msgq->buffer_end) {
+					msgq->read_ptr = msgq->buffer_start;
+				}
+			} else {
```
As a slot in the message queue buffer has been made available, I think that we can simplify (and simultaneously correct) the prepend logic. Instead of copying to `read_ptr` (which if I am following things correctly overwrites a message), we can copy to `write_ptr` and then revert `read_ptr` to that. 

```
(void)memcpy(msgq->write_ptr, pending_thread->base.swap_data, msgq->msg_size);
msgq->read_ptr = msgq->write_ptr;
```

TaiJuWu replied:
Interesting idea, but if the msg queue is empty and `thread A` and `thread B` call `msgq_put` both, and then `thread C` call `msgq_get` we will miss a item.
Is it right?

peter-mitsis replied:
If the message queue is empty, then we are not in this if-block where (msgq->used_msgs > 0U). Instead, we would be farther down in the else clause.

TaiJuWu replied:
Sorry, I think I need to explain more detail.
if the msg queue is empty and `thread A` and `thread B` call `msgq_put`.
There are two items in msgq.
`Thread C` call `msgq_get` it will enter is block where (msgq->used_msgs > 0U).
In this condition, `msgq->read_ptr = msgq->write_ptr` will miss an item.

peter-mitsis commented on include/zephyr/kernel.h:
```diff
@@ -4574,6 +4575,30 @@ int k_msgq_cleanup(struct k_msgq *msgq);
+ *
+ * @param msgq Address of the message queue.
+ * @param data Pointer to the message.
+ * @param timeout Non-negative waiting period to add the message,
```
I realize that this the timeout description was copy-pasted, but I think that description is out of date. Many of the other kernel APIs that use a timeout no longer use "Non-negative" as that is used to represent an absolute timeout.

I recommend dropping "Non-negative".

peter-mitsis commented on include/zephyr/kernel.h:
```diff
@@ -4574,6 +4575,30 @@ int k_msgq_cleanup(struct k_msgq *msgq);
+ * This routine sends a message to message queue @a q, inserting it before any
+ * other messages already in the queue.
+ *
+ * @note The message content is copied from @a data into @a msgq and the @a data
```
This should have a note or caveat clarifying the expected behaviour when the thread must pend because the message queue buffer is full as it is easy to miss/overlook.

If the message queue buffer is full, the calling thread may block. Only when it is unblocked can that message be inserted to the front of the message queue buffer. This means that the kernel can not guarantee LIFO ordering when the calling thread blocks.

TaiJuWu replied:
> This should have a note or caveat clarifying the expected behaviour when the thread must pend because the message queue buffer is full as it is easy to miss/overlook.
> 
> If the message queue buffer is full, the calling thread may block. Only when it is unblocked can that message be inserted to the front of the message queue buffer. This means that the kernel can not guarantee LIFO ordering when the calling thread blocks.

I think this scenario also guarantee LIFO because the calling thread prepend message to message queue.
And when new threads pending on a message queue, they also follows their sequence.

peter-mitsis replied:
LIFO can only be guaranteed within the message queue when there was space to add it to the message queue buffer. Consider the following ...

Let the message queue buffer have space for M messages. Let there be N threads already waiting to add messages into the message queue buffer. Let Thread X try to prepend a message but it is of lower priority than the N waiting threads; however it will be added to the end of the wait queue. If N >= M, then at least one of the threads ahead of Thread X in the wait queue will have its message added to and then retrieved from the message queue before Thread X can unpend have its message prepended. In this scenario, LIFO is not preserved.

TaiJuWu replied:
You are right. Maybe user need to know this information.

peter-mitsis reviewed (commented):
A few minor things, and one bug I think.

It might be worthwhile to update the documentation under doc/kernel/services/data_passing/message_queues.rst as well.

cfriedt commented on include/zephyr/kernel_structs.h:
```diff
@@ -80,6 +80,19 @@ extern "C" {
+#define _MSG_PUT (BIT(1))
+
+/* used to clear MSG_QUEUE pending reason */
+#define _MSG_QUEUE (_MSG_PREPEND & _MSG_PREPEND)
```
This doesn't look right - should this be a bitwise OR of the two constants above?

TaiJuWu replied:
Fix. Maybe it will remove. Because I find we don't need to bit to know the pending reason.

TaiJuWu commented:
Documentation improvment is on the way.

aescolar commented:
> And I am trying to fix CI.

 https://discord.com/channels/720317445772017664/1072528982278414448/1237616296208633876
> Is there someone have idea what reason make all bsim test fail on my new test case?

The code is segfaulting. In some platforms it is just lucky not to.
You can repro it easily locally. If you don't have babblesim, just build targeting native_sim with the address sanitizer
```
mkdir build && cd build
cmake -GNinja -DBOARD=native_sim ../tests/kernel/msgq/msgq_api/ -DCONFIG_ASAN=y
ninja
zephyr/zephyr.exe
```
```
...
==889310==ERROR: AddressSanitizer: global-buffer-overflow on address 0x08080f44 at pc 0xec8417bf bp 0xe73fd028 sp 0xe73fcbf8
WRITE of size 4 at 0x08080f44 thread T18
    #0 0xec8417be in __interceptor_memcpy ../../../../../src/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:827
    #1 0x8057594 in memcpy /usr/include/i386-linux-gnu/bits/string_fortified.h:29
    #2 0x8057594 in _msgq_put_internal zephyr/kernel/msg_q.c:171
    #3 0x804bec1 in k_msgq_prepend zephyr/build/zephyr/include/generated/syscalls/kernel.h:1207
    #4 0x804bec1 in msgq_api_1cpu_test_msgq_pending zephyr/tests/kernel/msgq/msgq_api/src/test_msgq_contexts.c:507
    #5 0x804bec1 in _msgq_api_1cpu_test_msgq_pending_wrapper zephyr/tests/kernel/msgq/msgq_api/src/test_msgq_contexts.c:490
    #6 0x80547f2 in run_test_functions zephyr/subsys/testsuite/ztest/src/ztest.c:315
    #7 0x80547f2 in test_cb zephyr/subsys/testsuite/ztest/src/ztest.c:649
    #8 0x804e9f5 in z_thread_entry zephyr/lib/os/thread_entry.c:48
...
0x08080f44 is located 60 bytes to the left of global variable 'tdata2' defined in 'zephyr/tests/kernel/msgq/msgq_api/src/test_msgq_contexts.c:19:17' (0x8080f80) of size 116
0x08080f44 is located 0 bytes to the right of global variable 'tbuffer1' defined in 'zephyr/tests/kernel/msgq/msgq_api/src/test_msgq_contexts.c:21:37' (0x8080f40) of size 4
...
```

TaiJuWu committed:
test: add a testcase to test message queue

Add test_msgq_pending to test message queue behavior when
the caller is going to prepend.

Signed-off-by: TaiJu Wu <tjwu1217@gmail.com>

TaiJuWu committed:
tracing: add msg queue prepend trace

For `k_msgq_prepend` add tracing feature.

Signed-off-by: TaiJu Wu <tjwu1217@gmail.com>

TaiJuWu committed:
doc: add message queue prepend document

Add document about k_msgq_prepend.

Signed-off-by: TaiJu Wu <tjwu1217@gmail.com>
